bin/e2hs-writer/src/reader.rs

```diff
@@ -0,0 +1,384 @@
+            .block
+            .message_capella()
+            .map_err(|e| anyhow!("Unable to decode capella block: {e:?}"))?;
+        let payload = block.body.execution_payload.clone();
+        let transactions = decode_transactions(&payload.transactions)?;
+        let header = pre_deneb_execution_payload_to_header(payload.clone(), &transactions)?;
+
+        let historical_batch = HistoricalBatch {
+            state_roots: era.era_state.state.state_roots().clone(),
+            block_roots: era.era_state.state.block_roots().clone(),
+        };
+        let slot = block.slot;
+        // beacon block proof
+        let hb_proof = historical_batch.build_block_root_proof(slot % EPOCH_SIZE);
+        let beacon_block_proof: FixedVector<B256, typenum::U13> = hb_proof.into();
+        // execution block proof
+        let mut execution_block_hash_proof = block.body.build_execution_block_hash_proof();
+        let body_root_proof = block.build_body_root_proof();
+        execution_block_hash_proof.extend(body_root_proof);
+        let execution_block_proof: VariableList<B256, typenum::U12> =
+            execution_block_hash_proof.into();
+        let proof = BlockProofHistoricalSummaries {
+            beacon_block_proof,
+            beacon_block_root: block.tree_hash_root(),
+            slot,
+            execution_block_proof,
+        };
+
+        let header_with_proof = HeaderWithProof {
+            header,
+            proof: BlockHeaderProof::HistoricalSummaries(proof),
+        };
+        let withdrawals: Vec<Withdrawal> =
+            payload.withdrawals.iter().map(Withdrawal::from).collect();
+        let body = BlockBody(AlloyBlockBody {
+            transactions,
+            ommers: vec![],
+            withdrawals: Some(Withdrawals::new(withdrawals)),
+        });
+        let receipts = self
+            .execution_api
+            .get_era_receipts(
+                block_number,
+                payload.transactions.len() as u64,
+                payload.receipts_root,
+            )
+            .await?;
+        Ok(AllBlockData {
+            block_number,
+            header_with_proof,
+            body,
+            receipts,
+        })
+    }
+
+    pub fn iter_blocks(self) -> impl Stream<Item = anyhow::Result<AllBlockData>> {
+        let mut current_block = self.starting_block;
+        stream! {
+        while current_block < self.ending_block {
+            if current_block < MERGE_BLOCK_NUMBER {
+                yield self.get_pre_merge_block_data(current_block);
+            } else if current_block < SHANGHAI_BLOCK_NUMBER {
+                yield self.get_pre_capella_block_data(current_block).await;
+            } else if current_block < CANCUN_BLOCK_NUMBER {
+                yield self.get_pre_deneb_block_data(current_block).await;
+            } else {
+                yield Err(anyhow!("Invalid block number: {current_block}"));
+            }
+            current_block += 1;
+        }
+        }
+    }
+}
```

Comment:
```suggestion
    pub fn iter_blocks(self) -> impl Stream<Item = anyhow::Result<AllBlockData>> {
        stream! {
            for current_block in (self.starting_block..self.ending_block) {
                if current_block < MERGE_BLOCK_NUMBER {
                    yield self.get_pre_merge_block_data(current_block);
                } else if current_block < SHANGHAI_BLOCK_NUMBER {
                    yield self.get_pre_capella_block_data(current_block).await;
                } else if current_block < CANCUN_BLOCK_NUMBER {
                    yield self.get_pre_deneb_block_data(current_block).await;
                } else {
                    yield Err(anyhow!("Invalid block number: {current_block}"));
                }
            }
        }
    }
}
```

It looks like indentation is missing here, also I think if we use an iterator the code would be a little simpler, then a while loop and maintaining a mut variable